### PR TITLE
Syntax Tree Visitor to AST

### PR DESCRIPTION
**Pull request by CasperKjaerhus**
Complete with an AST Visitor :)

### CODE REVIEW

File: Antlr/Test_Parser/Tests/Array_Usage.txt [hunk 1]
```diff
@@ -1,5 +1,7 @@
-﻿function main() returns void begin
-    number timer = 100;
+﻿include benis;
```
Comment (anesheim-uni) [inline Antlr/Test_Parser/Tests/Array_Usage.txt:1]: yeet

File: Antlr/Test_Parser/Tests/BoolsAndStrings.txt [hunk 2]
```diff
@@ -1,7 +1,7 @@
-﻿function main() returns void begin
+﻿function main(number penisLength) returns void begin
```
Comment (anesheim-uni) [inline Antlr/Test_Parser/Tests/BoolsAndStrings.txt:1]: main og penisLength wtf

File: Antlr/Test_Parser/Tests/Array_Usage.txt [hunk 3]
```diff
@@ -1,5 +1,7 @@
...
+
+function main() returns void begin
```
Comment (anesheim-uni) [inline Antlr/Test_Parser/Tests/Array_Usage.txt:3]: main

File: Antlr/Test_Parser/Tests/BoolsAndStrings.txt [hunk 4]
```diff
@@ -1,7 +1,7 @@
...
     string stringVar = "This is a string";
     
-    Serial.Print(getFirstChar(stringVar)); 
+    Serial.Print(getFirstChar(!stringVar));
```
Comment (anesheim-uni) [inline Antlr/Test_Parser/Tests/BoolsAndStrings.txt:4]: negation på string? hvad kommer der ud af dette?

File: Antlr/Test_Parser_src/Test_Parser/Tests/Array_Usage.txt [hunk 5]
```diff
@@ -1,5 +1,7 @@
-﻿function main() returns void begin
-    number timer = 100;
+﻿include benis;
+
+function main() returns void begin
```
Comment (anesheim-uni) [inline Antlr/Test_Parser_src/Test_Parser/Tests/Array_Usage.txt]: main

File: Antlr/Test_Parser_src/Test_Parser/Tests/includetest.txt [hunk 6]
```diff
@@ -0,0 +1,3 @@
+include asdf;
```
Comment (anesheim-uni) [inline Antlr/Test_Parser_src/Test_Parser/Tests/includetest.txt]: kom med real library/eller testLib

File: Shell/SALShell/SALShell/Parser/ASTVisitor.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+namespace SALShell.Parser
+{
+    abstract class ASTVisitor<T>
+    {
+        public virtual T Visit(ASTNode node)
+        {
+            foreach(ASTNode child in node.Children)
+            {
+                return Visit((dynamic) child);
```
Comment (anesheim-uni) [inline Shell/SALShell/SALShell/Parser/ASTVisitor.cs]: Consider comment for dynamic

File: Shell/SALShell/SALShell/Parser/ConcreteP4Visitor.cs [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+            if(context.True() != null)
+            {
+                return new ValueAstNode(context.True().Symbol);
+            }
+            else if(context.False() != null)
+            {
+                return new ValueAstNode(context.False().Symbol);
+            }
+
+            throw new Exception("Parse Error lulskibreski");
```
Comment (anesheim-uni) [inline Shell/SALShell/SALShell/Parser/ConcreteP4Visitor.cs]: fix

File: Shell/SALShell/SALShell/Parser/Nodes/AssignAstNode.cs [hunk 9]
```diff
@@ -0,0 +1,16 @@
+﻿using Antlr4.Runtime;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace SALShell.Parser
+{
+    public class AssignAstNode : ASTNode
+    {
+        public AssignAstNode(ASTNode id, IToken Assign, ASTNode Expr) : base(Assign)
```
Comment (anesheim-uni) [inline Shell/SALShell/SALShell/Parser/Nodes/AssignAstNode.cs]: casing

File: Shell/SALShell/SALShell/Parser/Nodes/ExprAstNode.cs [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace SALShell.Parser
+{
+    public class ExprAstNode : ASTNode
+    {
+        public bool Negation { get; set; }
+
+        public ExprAstNode(IToken Operator) : base(Operator)
```
Comment (anesheim-uni) [inline Shell/SALShell/SALShell/Parser/Nodes/ExprAstNode.cs]: casing

File: Shell/SALShell/SALShell/Parser/Nodes/IfStructureAstNode.cs [hunk 11]
```diff
@@ -0,0 +1,20 @@
+﻿using Antlr4.Runtime;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace SALShell.Parser
+{
+    class IfStructureAstNode : ASTNode
+    {
+        public IfStructureAstNode(ASTNode expr, ASTNode Body, ASTNode ElseStmt, IToken token) : base(token)
```
Comment (anesheim-uni) [inline Shell/SALShell/SALShell/Parser/Nodes/IfStructureAstNode.cs]: casing

File: Shell/SALShell/SALShell/Parser/Nodes/LogicAndAstNode.cs [hunk 12]
```diff
@@ -0,0 +1,16 @@
+﻿using Antlr4.Runtime;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace SALShell.Parser
+{
+    public class LogicAndAstNode : ExprAstNode
+    {
+        public LogicAndAstNode(ASTNode expr1, ASTNode expr2, IToken Operator) : base(Operator)
```
Comment (anesheim-uni) [inline Shell/SALShell/SALShell/Parser/Nodes/LogicAndAstNode.cs]: casing

File: Shell/SALShell/SALShell/Parser/Nodes/LogicEqualityAstNode.cs [hunk 13]
```diff
@@ -0,0 +1,16 @@
+﻿using Antlr4.Runtime;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace SALShell.Parser
+{
+    class LogicEqualityAstNode : ExprAstNode
+    {
+        public LogicEqualityAstNode(ASTNode expr1, ASTNode expr2, IToken Operator) : base(Operator)
```
Comment (anesheim-uni) [inline Shell/SALShell/SALShell/Parser/Nodes/LogicEqualityAstNode.cs]: casing

File: Shell/SALShell/SALShell/Parser/Nodes/LogicOrAstNode.cs [hunk 14]
```diff
@@ -0,0 +1,16 @@
+﻿using Antlr4.Runtime;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace SALShell.Parser
+{
+    class LogicOrAstNode : ExprAstNode
+    {
+        public LogicOrAstNode(ASTNode expr1, ASTNode expr2, IToken Operator) : base(Operator)
```
Comment (anesheim-uni) [inline Shell/SALShell/SALShell/Parser/Nodes/LogicOrAstNode.cs]: casing

File: Shell/SALShell/SALShell/Parser/Nodes/MultAstNode.cs [hunk 15]
```diff
@@ -0,0 +1,13 @@
+﻿using Antlr4.Runtime;
+
+namespace SALShell.Parser
+{
+    public class MultAstNode : ExprAstNode
+    {
+        public MultAstNode(ASTNode expr1, ASTNode expr2, IToken Operator) : base(Operator)
```
Comment (anesheim-uni) [inline Shell/SALShell/SALShell/Parser/Nodes/MultAstNode.cs]: casing

File: Shell/SALShell/SALShell/Parser/Nodes/PostfixExprAstNode.cs [hunk 16]
```diff
@@ -0,0 +1,15 @@
+﻿using Antlr4.Runtime;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace SALShell.Parser
+{
+    class PostfixExprAstNode : ExprAstNode
+    {
+        public PostfixExprAstNode(ASTNode idnode, IToken Operator) : base(Operator)
```
Comment (anesheim-uni) [inline Shell/SALShell/SALShell/Parser/Nodes/PostfixExprAstNode.cs]: casing

File: Shell/SALShell/SALShell/Parser/Nodes/PrefixExprAstNode.cs [hunk 17]
```diff
@@ -0,0 +1,15 @@
+﻿using Antlr4.Runtime;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace SALShell.Parser
+{
+    class PrefixExprAstNode : ExprAstNode
+    {
+        public PrefixExprAstNode(ASTNode idnode, IToken Operator) : base(Operator)
```
Comment (anesheim-uni) [inline Shell/SALShell/SALShell/Parser/Nodes/PrefixExprAstNode.cs]: casing

File: Shell/SALShell/SALShell/Parser/Nodes/RelationalExprAstNode.cs [hunk 18]
```diff
@@ -0,0 +1,16 @@
+﻿using Antlr4.Runtime;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace SALShell.Parser
+{
+    class RelationalExprAstNode : ExprAstNode
+    {
+        public RelationalExprAstNode(ASTNode expr1, ASTNode expr2, IToken Operator) : base(Operator)
```
Comment (anesheim-uni) [inline Shell/SALShell/SALShell/Parser/Nodes/RelationalExprAstNode.cs]: casing

File: Shell/SALShell/SALShell/Parser/Nodes/StatementAstNode.cs [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+    class StatementAstNode : ASTNode
+    {
+        public StatementAstNode NextNode { get; private set; }
+        public ASTNode Action { get; }
+        public StatementAstNode(ASTNode action, IToken token) : base(token)
+        {
+            this.Action = action;
+        }
+
+        public void Push(StatementAstNode item)
```
Comment (anesheim-uni) [inline Shell/SALShell/SALShell/Parser/Nodes/StatementAstNode.cs:17]: This is driver for hele ast generation right?

File: Shell/SALShell/SALShell/Parser/Nodes/SwitchItemAstNode.cs [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+using System.Text;
+
+namespace SALShell.Parser
+{
+    public class SwitchItemAstNode : ASTNode
+    {
+        public ASTNode Value { get; }
+        public ASTNode Block { get; }
+        public bool IsBreaked { get; }
+        public SwitchItemAstNode(ASTNode value, ASTNode block, bool isBreaked, IToken token) : base(token)
```
Comment (anesheim-uni) [inline Shell/SALShell/SALShell/Parser/Nodes/SwitchItemAstNode.cs:13]: breaked yuk xd

File: Shell/SALShell/SALShell/Program.cs [hunk 21]
```diff
@@ -9,10 +13,28 @@ class Program
...
-            ICore core = new Core.Core();
-            IShellUI ui = new ShellCLI();
-            ShellController sheeController = new ShellController(ui, core);
-            sheeController.Ui.Start();
+            //ICore core = new Core.Core();
+            //IShellUI ui = new ShellCLI();
+            //ShellController sheeController = new ShellController(ui, core);
+            //sheeController.Ui.Start();
+
+            string text = System.IO.File.ReadAllText(@"D:\Github Repos\SAL\Antlr\Test_Parser_src\Test_Parser\Tests\BoolsAndStrings.txt");
```
Comment (anesheim-uni) [inline Shell/SALShell/SALShell/Program.cs:21]: hardcoded only work for u?

File: Shell/SALShell/SALShell/Program.cs [hunk 22]
```diff
@@ -9,10 +13,28 @@ class Program
     {
         static void Main(string[] args)
```
Comment (anesheim-uni) [inline Shell/SALShell/SALShell/Program.cs:14]: Hvordan er shell'en incorporated med resten pt?

File: Shell/SALShell/SALShell/Parser/Nodes/SwitchItemAstNode.cs (hunk 20 above)
Comment (CasperKjaerhus) [reply Shell/SALShell/SALShell/Parser/Nodes/SwitchItemAstNode.cs:13]: IS BROKE?

File: Shell/SALShell/SALShell/Program.cs (hunk 21 above)
Comment (CasperKjaerhus) [reply Shell/SALShell/SALShell/Program.cs:21]: Yarp

**Review by anesheim-uni — COMMENTED** [review]
Check comments my dude

File: Shell/SALShell/SALShell/Program.cs (hunk 22 above)
Comment (CasperKjaerhus) [reply Shell/SALShell/SALShell/Program.cs:14]: Overhovedet ikke

**Pull request merged**